### DISCUSSION TITLE
SDNN can take both spikes and image as input?

### DISCUSSION THREAD

**Discussion by tangores** [lava-nc/lava-dl#225] [General]
I was wondering what kind of input can sigma-delta neural networks take, they can take both spikes & images as input?

**Comment by bamsumit** [comment]
Yes, why not. I assume you mean binary spikes. You will not need the delay layer in the input that encodes images into graded spikes in this case.

**Reply by tangores** [reply]
Is there a way to convert the input into PSP (just like Slayer had for spiking networks)?

**Comment by tangores** [comment]
Is there a way to convert the input into PSP for SDNNs (just like Slayer had for spiking networks)?

**Reply by bamsumit** [reply]
Yes you can. There are multiple ways to achieve it.
* You can apply any FIR filter as PSP response using `slayer.utils.conv` or `slayer.utils.FIR`. There is a FIR filter bank module as well. These are similar to SLAYER PyTorch.
* If you want exponentially decaying PSP, you can apply leaky integrator dynamics
    ```python
    psp0 = slayer.neuron.dynamics.leaky_integrator.dynamics(input, decay=decay, state=initial_state, w_scale=scale)  # Note there is no threshold parameter supplied here
    ```
* You can apply it twice to get the alpha PSP response
    ```python
    psp1 = slayer.neuron.dynamics.leaky_integrator.dynamics(psp, decay=decay, state=initial_state, w_scale=scale)  # Note there is no threshold parameter supplied here
    ```
* Similarly you can apply other dynamics like resonator dyanamics. Look into `slayer.neuron.dynamics`.

**Comment by Kristi1217** [comment]
Hi @bamsumit @tangores ,

I tried using psp0 to convert the inputs into binary spikes then the psp0 is sent SDNN layers.  
1) Is there a way to verify that the input was converted? because I don't see much difference while training a network, if it takes in normal channels or binary spikes, it computes at the same rate. 

2) My forward operations looks something like this. Secondly I cant change number of output channels, I have to then to create separate blocks. 
```

      self.blocks = torch.nn.ModuleList([
          slayer.block.sigma_delta.Conv(sdnn_params1,  2, 2, 3, padding=1),            
          slayer.block.sigma_delta.Conv(sdnn_params1,  2, 2, 1, padding=0), 
      ])
     
      
  def forward(self, x):
      count = []
      event_cost = 0
      device = torch.device("cuda")  # or "cpu" if you want to use the CPU

      scale = 1#<<12 # scale factor for integer simulation
      decay = torch.FloatTensor([0.1 * scale]).to(device)
      initial_state = torch.FloatTensor([0]).to(device)
      threshold = 0.5
      
      B, C, H, W, T = x.shape
      
      psp = slayer.neuron.dynamics.leaky_integrator._li_dynamics_fwd(x, decay=decay, state=initial_state, w_scale=scale,threshold= threshold)  
      
         
      for block in self.blocks:
          x = block(psp)
          if hasattr(block, 'neuron'):
              event_cost += event_rate_loss(x)
              count.append(torch.sum(torch.abs((x[..., 1:]) > 0).to(x.dtype)).item())
              
      return x, event_cost, torch.FloatTensor(count).reshape((1, -1)).to(x.device)
```
